vhdx/src/vhdx_metadata.rs

```diff
@@ -0,0 +1,285 @@
+    // Not a metadata sign
+    InvalidMetadataSign,
+    // Physical sector size doesn't meet spec
+    InvalidPhysicalSectorSize,
+    // Not all required metadata are present
+    MissingMetadata,
+    // Failed to read metadata from file
+    ReadMetadata(io::Error),
+    // Found a flag not supported by this implementation
+    UnsupportedFlag,
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+impl Display for Error {
+    #[remain::check]
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        #[sorted]
+        match self {
+            InvalidBlockSize => write!(f, "invalid block size count"),
+            InvalidEntryCount => write!(f, "invalid metadata entry count"),
+            InvalidLogicalSectorSize => write!(f, "invalid logical sector size"),
+            InvalidMetadataItem => write!(f, "invalid metadata ID"),
+            InvalidMetadataSign => write!(f, "metadata sign doesn't match"),
+            InvalidPhysicalSectorSize => write!(f, "invalid logical sector size"),
+            MissingMetadata => write!(f, "not all required metadata found"),
+            ReadMetadata(e) => write!(f, "failed to read metadata headers: {}", e),
+            UnsupportedFlag => write!(f, "this implementation support this metadata flag"),
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct DiskSpec {
+    disk_id: u128,
+    block_size: u32,
+    sectors_per_block: u32,
+    virtual_disk_size: u64,
+    logical_sector_size: u32,
+    physical_sector_size: u32,
+    chunk_ratio: u64,
+}
+
+struct MetadataTableHeader {
+    signature: u64,
+    reserved: u16,
+    entry_count: u16,
+    reserved2: [u8; 20],
+}
+
+impl MetadataTableHeader {
+    pub fn new(buffer: &[u8]) -> Result<MetadataTableHeader> {
+        let signature = LittleEndian::read_u64(&buffer[0..8]);
+        if signature != METADATA_SIGN {
+            return Err(Error::InvalidMetadataSign);
+        }
+
+        let reserved = LittleEndian::read_u16(&buffer[8..10]);
+        let entry_count = LittleEndian::read_u16(&buffer[10..12]);
+        if entry_count > METADATA_MAX_ENTRIES {
+            return Err(Error::InvalidEntryCount);
+        }
+
+        Ok(MetadataTableHeader {
+            signature,
+            reserved,
+            entry_count,
+            reserved2: buffer[12..32]
+                .try_into()
+                .expect("slice with incorrect length"),
+        })
+    }
+}
+
+pub struct MetadataTableEntry {
+    item_id: u128,
+    offset: u32,
+    length: u32,
+    flag_bits: u32,
+    reserved: u32,
+}
+
+impl MetadataTableEntry {
+    fn new(buffer: &[u8]) -> Result<MetadataTableEntry> {
+        let item_id = LittleEndian::read_u128(&buffer[0..16]);
+        let offset = LittleEndian::read_u32(&buffer[16..20]);
+        let length = LittleEndian::read_u32(&buffer[20..24]);
+        let flag_bits = LittleEndian::read_u32(&buffer[24..28]);
+        let reserved = LittleEndian::read_u32(&buffer[28..32]);
```

Comment:
Simplify by filling the structure directly.

Reply:
Done.